File: csharp/EgyNames/Corrector.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace EgyptianNames
{
    public static class Corrector
    {
        public static string CorrectToken(string token, string? dataPath = null)
        {
            if (string.IsNullOrWhiteSpace(token)) return token;
            string t = token.Trim();

            // 1. Direct surface correction pair
            string? canonical = LookupIndices.GetCorrection(t, dataPath);
            if (canonical != null) return canonical;

            // 2. Exact match in arabic index (including phonetic variants)
            var entry = LookupIndices.LookupAr(t, dataPath);
            if (entry != null) return entry.Ar;

            // 3. Normalized form lookup
            string norm = LookupIndices.NormalizeAr(t);
            var arNorm = LookupIndices.GetArNormForms(dataPath);
            if (arNorm.TryGetValue(norm, out var normEntry)) return normEntry.Ar;

            // 4. Trailing Alif / Alif Maqsura check
            if (norm.EndsWith("\u0627"))
            {
                string alt = norm.Substring(0, norm.Length - 1) + "\u064A";
                if (arNorm.TryGetValue(alt, out var altMatch)) return altMatch.Ar;
            }
            else if (norm.EndsWith("\u064A"))
            {
                string alt = norm.Substring(0, norm.Length - 1) + "\u0627";
                if (arNorm.TryGetValue(alt, out var altMatch)) return altMatch.Ar;
            }

            return t;
        }

        public static string Correct(string name, string? dataPath = null)
        {
            if (string.IsNullOrWhiteSpace(name)) return name;
            var tokens = name.Trim().Split(new[] { ' ' }, StringSplitOptions.RemoveEmptyEntries);
            var result = new List<string>();

            for (int i = 0; i < tokens.Length; i++)
            {
                string current = tokens[i];

                // Check compound pair (e.g. "عبد" + "الرحيم" -> "عبدالرحيم")
                if (i < tokens.Length - 1)
                {
                    string next = tokens[i + 1];
                    string compound = $"{current} {next}";
                    string compoundNoSpace = $"{current}{next}";

                    var compoundEntry = LookupIndices.LookupAr(compound, dataPath) ?? LookupIndices.LookupAr(compoundNoSpace, dataPath);
                    if (compoundEntry != null)
                    {
                        result.Add(compoundEntry.Ar);
                        i++; // skip second part
                        continue;
                    }
                }

                result.Add(CorrectToken(current, dataPath));
            }

            return string.Join(" ", result);
        }
    }
}

File: showcase/src/engine/corrector.ts
import {
  getArNormForms,
  getCorrection,
  lookupAr,
  normalizeAr,
} from "./lookupIndices";

export function correctToken(token: string): string {
  if (!token || !token.trim()) return token;
  const t = token.trim();

  // 1. Direct surface correction pair
  const canonical = getCorrection(t);
  if (canonical) return canonical;

  // 2. Exact match in arabic index (including phonetic variants)
  const exact = lookupAr(t);
  if (exact) return exact.ar;

  // 3. Normalized form lookup
  const norm = normalizeAr(t);
  const arNorm = getArNormForms();
  const normMatch = arNorm.get(norm);
  if (normMatch) return normMatch.ar;

  // 4. Trailing Alif / Alif Maqsura check
  if (norm.endsWith("\u0627")) {
    const alt = norm.slice(0, -1) + "\u064A";
    const altMatch = arNorm.get(alt);
    if (altMatch) return altMatch.ar;
  } else if (norm.endsWith("\u064A")) {
    const alt = norm.slice(0, -1) + "\u0627";
    const altMatch = arNorm.get(alt);
    if (altMatch) return altMatch.ar;
  }

  return t;
}

export function correct(name: string): string {
  if (!name || !name.trim()) return name;
  const rawTokens = name.trim().split(/\s+/);
  const result: string[] = [];

  for (let i = 0; i < rawTokens.length; i++) {
    const current = rawTokens[i];

    // Check if current token forms a compound name with next token (e.g. "عبد" + "الرحيم" -> "عبدالرحيم" or "أبو" + "بكر")
    if (i < rawTokens.length - 1) {
      const next = rawTokens[i + 1];
      const compound = `${current} ${next}`;
      const compoundNoSpace = `${current}${next}`;

      const compoundEntry = lookupAr(compound) || lookupAr(compoundNoSpace);
      if (compoundEntry) {
        result.push(compoundEntry.ar);
        i++; // skip compound second part
        continue;
      }
    }

    result.push(correctToken(current));
  }

  return result.join(" ");
}
